File: node/examples/parent_child_struct_monitor.ts
const childProcess = require("node:child_process");
const timers = require("node:timers/promises");

const xproc = require("../index.js") as XprocModule;

type TelemetryPacket = {
  message: string;
  a: number;
  b: number;
};

type CliOptions = {
  child: boolean;
  shmPath: string | null;
  ticks: number;
  intervalMs: number;
};

const CHILD_FLAG = "--pc-struct-child";
const MESSAGE_BYTES = 256;
const INT32_BYTES = 4;
const PACKET_SIZE = MESSAGE_BYTES + (INT32_BYTES * 2);
const DATA_CAPACITY = 32768n;
const POLL_INTERVAL_MS = 100;

const delay = timers.setTimeout as (ms: number) => Promise<void>;

function parseCli(argv: readonly string[]): CliOptions {
  let child = false;
  let shmPath: string | null = null;
  let ticks = 100;
  let intervalMs = 300;

  for (let i = 0; i < argv.length; ++i) {
    const arg = argv[i];
    if (arg === CHILD_FLAG) {
      child = true;
      shmPath = argv[i + 1] ?? null;
      i += 1;
      continue;
    }
    if (arg.startsWith("--ticks=")) {
      ticks = parseIntegerFlag(arg.slice("--ticks=".length), "ticks");
      continue;
    }
    if (arg.startsWith("--interval-ms=")) {
      intervalMs = parseIntegerFlag(arg.slice("--interval-ms=".length), "interval-ms");
      continue;
    }
  }

  return { child, shmPath, ticks, intervalMs };
}

function parseIntegerFlag(raw: string, name: string): number {
  const value = Number.parseInt(raw, 10);
  if (!Number.isFinite(value) || value < 0) {
    throw new Error(`invalid --${name} value: ${raw}`);
  }
  return value;
}

function createProducerOptions(path: string): TransportOptions {
  return {
    path,
    shmSize: xproc.XPROC_C_INFER_EXISTING_SHM_SIZE,
    channelType: "fixed",
    itemSize: PACKET_SIZE,
    createIfMissing: false,
  };
}

function createConsumerOptions(path: string): TransportOptions {
  return {
    path,
    shmSize: xproc.shmSizeForDataCapacity(DATA_CAPACITY),
    channelType: "fixed",
    itemSize: PACKET_SIZE,
    createIfMissing: true,
  };
}

function encodePacket(packet: TelemetryPacket): Buffer {
  const buffer = Buffer.alloc(PACKET_SIZE);
  const encodedMessage = Buffer.from(packet.message, "utf8");
  encodedMessage.copy(buffer, 0, 0, Math.min(encodedMessage.length, MESSAGE_BYTES - 1));
  buffer.writeInt32LE(packet.a, MESSAGE_BYTES);
  buffer.writeInt32LE(packet.b, MESSAGE_BYTES + INT32_BYTES);
  return buffer;
}

function decodePacket(payload: Uint8Array): TelemetryPacket {
  const buffer = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
  let messageEnd = 0;
  while (messageEnd < MESSAGE_BYTES && buffer[messageEnd] !== 0) {
    messageEnd += 1;
  }

  return {
    message: buffer.toString("utf8", 0, messageEnd),
    a: buffer.readInt32LE(MESSAGE_BYTES),
    b: buffer.readInt32LE(MESSAGE_BYTES + INT32_BYTES),
  };
}

function cleanupShm(path: string): void {
  try {
    xproc.shmUnlink(path);
  } catch {
    // Best-effort cleanup only.
  }
}

async function runChildWriter(path: string, ticks: number, intervalMs: number): Promise<void> {
  const producer = new xproc.Producer(createProducerOptions(path));
  try {
    for (let i = 0; i <= ticks; ++i) {
      const packet = encodePacket({
        message: `tick-${i}`,
        a: i,
        b: i * 2,
      });
      producer.sendFixedSized(packet);
      await delay(intervalMs);
    }
  } finally {
    producer.close();
  }
}

function drainConsumer(consumer: Consumer, onPacket: (payload: Uint8Array) => void): boolean {
  let consumedAny = false;
  while (true) {
    const payload = consumer.pollCopy();
    if (payload === null) {
      return consumedAny;
    }
    consumedAny = true;
    if (payload.length !== PACKET_SIZE) {
      continue;
    }
    onPacket(payload);
  }
}

async function runParent(ticks: number, intervalMs: number): Promise<void> {
  const shmPath = `/xproc_example_parent_child_struct_${process.pid}`;
  cleanupShm(shmPath);

  const consumer = new xproc.Consumer(createConsumerOptions(shmPath));
  let childExitCode: number | null = null;
  let lastPacketBytes: Buffer | null = null;

  const child = childProcess.spawn(
    process.execPath,
    ["--experimental-strip-types", process.argv[1], CHILD_FLAG, shmPath, `--ticks=${ticks}`, `--interval-ms=${intervalMs}`],
    {
      stdio: "inherit",
    },
  );

  child.on("exit", (code: number | null) => {
    childExitCode = code ?? 1;
  });

  try {
    while (childExitCode === null) {
      drainConsumer(consumer, (payload) => {
        const currentBytes = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
        if (lastPacketBytes !== null && Buffer.compare(currentBytes, lastPacketBytes) === 0) {
          return;
        }

        const packet = decodePacket(currentBytes);
        console.log(`message=${packet.message}, a=${packet.a}, b=${packet.b}`);
        lastPacketBytes = Buffer.from(currentBytes);
      });
      await delay(POLL_INTERVAL_MS);
    }

    drainConsumer(consumer, (payload) => {
      const currentBytes = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
      if (lastPacketBytes !== null && Buffer.compare(currentBytes, lastPacketBytes) === 0) {
        return;
      }

      const packet = decodePacket(currentBytes);
      console.log(`message=${packet.message}, a=${packet.a}, b=${packet.b}`);
      lastPacketBytes = Buffer.from(currentBytes);
    });
  } finally {
    consumer.close();
    cleanupShm(shmPath);
  }

  if (childExitCode !== 0) {
    throw new Error("child process failed");
  }
  console.log("child exited, parent done");
}

async function main(): Promise<void> {
  const cli = parseCli(process.argv.slice(2));
  if (cli.child) {
    if (cli.shmPath === null) {
      throw new Error(`${CHILD_FLAG} requires a shared-memory path`);
    }
    await runChildWriter(cli.shmPath, cli.ticks, cli.intervalMs);
    return;
  }

  await runParent(cli.ticks, cli.intervalMs);
}

void main().catch((error: unknown) => {
  const message = error instanceof Error ? error.stack ?? error.message : String(error);
  console.error(message);
  process.exit(1);
});

File: Python/examples/parent_child_struct_monitor.py
from __future__ import annotations

import argparse
import os
import struct
import subprocess
import sys
import time

from _common import cleanup_shm, load_xproc, module_dir_for_child, non_negative_int


CHILD_FLAG = "--pc-struct-child"
MESSAGE_BYTES = 256
INT32_BYTES = 4
PACKET_SIZE = MESSAGE_BYTES + (INT32_BYTES * 2)
DATA_CAPACITY = 32_768
POLL_INTERVAL_MS = 100


def create_producer_options(xproc, shm_path: str):
    options = xproc.TransportOptions()
    options.path = shm_path
    options.shm_size = xproc.INFER_EXISTING_SHM_SIZE
    options.channel_type = xproc.ChannelType.FIXED
    options.item_size = PACKET_SIZE
    options.create_if_missing = False
    return options


def create_consumer_options(xproc, shm_path: str):
    options = xproc.TransportOptions()
    options.path = shm_path
    options.shm_size = xproc.shm_size_for_data_capacity(DATA_CAPACITY)
    options.channel_type = xproc.ChannelType.FIXED
    options.item_size = PACKET_SIZE
    options.create_if_missing = True
    return options


def encode_packet(message: str, a: int, b: int) -> bytes:
    packet = bytearray(PACKET_SIZE)
    encoded_message = message.encode("utf-8")
    packet[: min(len(encoded_message), MESSAGE_BYTES - 1)] = encoded_message[: MESSAGE_BYTES - 1]
    struct.pack_into("<ii", packet, MESSAGE_BYTES, a, b)
    return bytes(packet)


def decode_packet(payload: bytes) -> tuple[str, int, int]:
    message_end = payload.find(b"\0", 0, MESSAGE_BYTES)
    if message_end < 0:
        message_end = MESSAGE_BYTES

    message = payload[:message_end].decode("utf-8")
    a, b = struct.unpack_from("<ii", payload, MESSAGE_BYTES)
    return message, a, b


def drain_consumer(consumer, on_packet) -> None:
    while True:
        payload = consumer.poll_copy()
        if payload is None:
            return
        if len(payload) != PACKET_SIZE:
            continue
        on_packet(payload)


def run_child_writer(xproc, shm_path: str, ticks: int, interval_ms: int) -> None:
    producer = xproc.Producer(create_producer_options(xproc, shm_path))
    try:
        for index in range(ticks + 1):
            producer.send_fixed_sized(encode_packet(f"tick-{index}", index, index * 2))
            time.sleep(interval_ms / 1000.0)
    finally:
        producer.close()


def run_parent(xproc, ticks: int, interval_ms: int) -> None:
    shm_path = f"/xproc_example_parent_child_struct_{os.getpid()}"
    cleanup_shm(xproc, shm_path)

    consumer = xproc.Consumer(create_consumer_options(xproc, shm_path))
    last_packet_bytes: bytes | None = None

    child_argv = [
        sys.executable,
        str(os.path.abspath(__file__)),
        CHILD_FLAG,
        shm_path,
        f"--ticks={ticks}",
        f"--interval-ms={interval_ms}",
    ]
    child_module_dir = module_dir_for_child(xproc)
    if child_module_dir is not None:
        child_argv.extend(["--module-dir", child_module_dir])

    child = subprocess.Popen(child_argv)

    try:
        while child.poll() is None:
            def print_packet(payload: bytes) -> None:
                nonlocal last_packet_bytes
                if last_packet_bytes == payload:
                    return

                message, a, b = decode_packet(payload)
                print(f"message={message}, a={a}, b={b}")
                last_packet_bytes = bytes(payload)

            drain_consumer(consumer, print_packet)
            time.sleep(POLL_INTERVAL_MS / 1000.0)

        def print_packet(payload: bytes) -> None:
            nonlocal last_packet_bytes
            if last_packet_bytes == payload:
                return

            message, a, b = decode_packet(payload)
            print(f"message={message}, a={a}, b={b}")
            last_packet_bytes = bytes(payload)

        drain_consumer(consumer, print_packet)
    finally:
        consumer.close()
        cleanup_shm(xproc, shm_path)

    if child.returncode != 0:
        raise RuntimeError("child process failed")

    print("child exited, parent done")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(CHILD_FLAG, dest="child_shm_path")
    parser.add_argument("--module-dir")
    parser.add_argument("--ticks", type=non_negative_int, default=100)
    parser.add_argument("--interval-ms", type=non_negative_int, default=300)
    args = parser.parse_args()

    xproc = load_xproc(args.module_dir)

    if args.child_shm_path is not None:
        run_child_writer(xproc, args.child_shm_path, args.ticks, args.interval_ms)
        return 0

    run_parent(xproc, args.ticks, args.interval_ms)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
